File: Cypher.cpp
#include<bits/stdc++.h>
using namespace std;
string alp="abcdefghijklmnopqrstuvwxyz";
void make_matrix(string* mat){
    for(int i=0;i<26;i++){
        mat[i]=alp.substr(alp.length()-i,alp.length())+alp.substr(0,alp.length()-i);
    }
    //for(int i=0;i<26;i++){
    //    cout<<mat[i]<<endl;
    //}
}

void encrypt_mssg(string key, string mssg, string* mat){
    string encrypt="";
    for(int i=0;i<mssg.length();i++){
        if(mssg[i]>96 && mssg[i]<=122){
            int a=0,b=0;
            for(int j=0;j<26;j++){
                if(mat[j][0]==key[i%key.length()]){
                    a=j;
                }
                if(mat[0][j]==mssg[i]){
                    b=j;
                }
            }
            encrypt+=mat[a][b];
        } else {
            encrypt+=mssg[i];
        }
    }
    cout<<encrypt<<endl;
}
void decrypt_mssg(string key, string* mat){
    string mssg;
    cout<<"Enter encrypted message: ";
    cin.ignore();
    getline(cin,mssg);
    string decrypt="";
    for(int i=0;i<mssg.length();i++){
        if(mssg[i]>96 && mssg[i]<=122){
            int a=0,b=0;
            for(int j=0;j<26;j++){
                if(mat[j][0]==key[i%key.length()]){
                    a=j;
                    for(int k=0;k<26;k++){
                        if(mat[a][k]==mssg[i]){
                            b=k;
                        }
                    }
                }
            }
            decrypt+=mat[0][b];
        } else {
            decrypt+=mssg[i];
        }
    }
    cout<<decrypt<<endl;
}
int main(){
    string matrix[26];
    make_matrix(matrix);
    string key="a";
    string mssg="hello";
    cout<<"------Substitution Cipher-------"<<endl;
    int choice=0;
    while(true){
        cout<<"1)New message"<<endl;
        cout<<"2)New key"<<endl;
        cout<<"3)View Message"<<endl;
        cout<<"4)Get decrypted message"<<endl;
        cout<<"5)Exit"<<endl;
        cout<<"Enter choice:";
        cin>>choice;
        cout<<endl;
        switch(choice){
            case 1:
                cout<<"Enter message:";
                cin.ignore();
                getline(cin,mssg);
                break;
            case 2:
                cout<<"Enter key:";
                cin.ignore();
                getline(cin,key);
                break;
            case 3:
                encrypt_mssg(key,mssg,matrix);
                break;
            case 4:
                decrypt_mssg(key,matrix);
                break;
            case 5:
                exit(0);
            default:
                cout<<"Invalid choice"<<endl;
        }
        cout<<endl;
    }
}

File: cypher.py
import string

alpl = string.ascii_lowercase
alph = string.ascii_uppercase

# def make_matrix():
#     mat = []
#     for i in range(26):
#         mat.append(alp[-i:] + alp[:-i])
#     return mat

def encrypt_mssg(key, mssg):
    encrypt = ""
    matl = []
    for i in range(26):
        matl.append(alpl[-i:] + alpl[:-i])

    for i in range(len(mssg)):
        if 'a' <= mssg[i] <= 'z':
            a = b = 0
            for j in range(26):
                if matl[j][0] == key[i % len(key)]:
                    a = j
                if matl[0][j] == mssg[i]:
                    b = j
            encrypt += matl[a][b]
        else:
            encrypt += mssg[i]
    return encrypt

def decrypt_mssg(key,mssg):

    decrypt = ""
    matl = []
    for i in range(26):
        matl.append(alpl[-i:] + alpl[:-i])
    for i in range(len(mssg)):

        if 'a' <= mssg[i] <= 'z':
            a = b = 0
            for j in range(26):
                if matl[j][0] == key[i % len(key)]:
                    a = j
                    for k in range(26):
                        if matl[a][k] == mssg[i]:
                            b = k
            decrypt += matl[0][b]
        else:
            decrypt += mssg[i]
    return decrypt

def main():

    key = "a"
    mssg = "hello"
    print("------Substitution Cipher-------")
    while True:
        print("1) New message")
        print("2) New key")
        print("3) View Message")
        print("4) Get decrypted message")
        print("5) Exit")
        choice = int(input("Enter choice: "))
        print()
        if choice == 1:
            mssg = input("Enter message: ")
        elif choice == 2:
            key = input("Enter key: ")
        elif choice == 3:
            encrypt_mssg(key, mssg, matrix)
        elif choice == 4:
            decrypt_mssg(key, matrix)
        elif choice == 5:
            break
        else:
            print("Invalid choice")
        print()

if __name__ == "__main__":
    main()